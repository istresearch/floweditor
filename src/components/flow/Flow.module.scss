@import 'variables.module.scss';

.empty_flow {
  margin: 30px;
  width: 400px;

  h1 {
    font-family: 'Roboto', sans-serif;
    letter-spacing: 1px;
    padding: 0;
    text-transform: none;
    font-weight: 100;
    font-size: 18px;
    margin: 0;
  }

  div {
    margin-bottom: 20px;
    font-size: 14px;
    line-height: 18px;
  }
}