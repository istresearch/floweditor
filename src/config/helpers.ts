import {
  FlowTypes,
  FlowTypeVisibility,
  Operator,
  Type,
  VISIBILITY_ONLINE
} from 'config/interfaces';
import { FlowEditorConfig } from 'flowTypes';

export const isOnlineFlowType = (flowType: FlowTypes) => {
  return !!VISIBILITY_ONLINE.find((type: FlowTypes) => type === flowType);
};

export const filterOperators = (operators: Operator[], config: FlowEditorConfig): Operator[] => {
  return filterVisibility(operators, config);
};

export const filterTypeConfigs = (typeConfigs: Type[], config: FlowEditorConfig): Type[] => {
  return filterVisibility(typeConfigs, config);
};

const filterVisibility = (items: FlowTypeVisibility[], config: FlowEditorConfig): any[] => {
  return items.filter((item: FlowTypeVisibility) => {
    // if we have a filter on our type, don't return it unless its present in our config
    if (item.filter) {
      if (!(config.filters || []).find((name: string) => name === item.filter)) {
        return false;
      }
    }

    if (item.visibility === undefined) {
      return true;
    }

    return item.visibility.findIndex((ft: FlowTypes) => ft === config.flowType) > -1;
  });
};
