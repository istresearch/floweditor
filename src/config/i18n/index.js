import i18n from 'i18next';
import { initReactI18next } from 'react-i18next';
import LanguageDetector from 'i18next-browser-languagedetector';

// TODO: consider chunking out translation files
// and using something like XHR backend to lazily
// load only the required language.
// See https://www.i18next.com/how-to/add-or-load-translations

import defaults from './defaults.json';
import es from './es/resource.json';
import fr from './fr/resource.json';
import ptBR from './pt-br/resource.json';

i18n
  .use(LanguageDetector)
  .use(initReactI18next)
  .init({
    detection: {
      order: ['querystring', 'htmlTag']
    },
    resources: {
      defaults: { translation: defaults },
      es: { translation: es },
      fr: { translation: fr },
      'pt-br': { translation: ptBR }
    },
    fallbackLng: ['defaults'],
    interpolation: {
      prefix: '[[',
      suffix: ']]'
    },
    lowerCaseLng: true
  });

export default i18n;
