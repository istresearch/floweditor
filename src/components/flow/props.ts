import { Type } from 'config/interfaces';
import { AnyAction, ContactProperties, FlowIssue } from 'flowTypes';
import { Asset, AssetStore, AssetType, RenderNode } from 'store/flowContext';
import { NodeEditorSettings } from 'store/nodeEditor';
import { DispatchWithState, GetState } from 'store/thunks';
import { titleCase } from 'utils';

export interface IssueProps {
  helpArticles: { [key: string]: string };
  issues: FlowIssue[];
}

export interface ActionFormProps extends IssueProps {
  // action details
  nodeSettings: NodeEditorSettings;
  typeConfig: Type;
  assetStore: AssetStore;

  addAsset(assetType: string, asset: Asset): void;

  // update handlers
  updateAction(
    action: AnyAction,
    onUpdated?: (dispatch: DispatchWithState, getState: GetState) => void
  ): void;

  // modal notifiers
  onTypeChange(config: Type): void;
  onClose(canceled: boolean): void;
}

export interface RouterFormProps extends IssueProps {
  nodeSettings: NodeEditorSettings;
  typeConfig: Type;

  assetStore: AssetStore;

  // update handlers
  updateRouter(renderNode: RenderNode): void;

  // modal notifiers
  onTypeChange(config: Type): void;
  onClose(canceled: boolean): void;
}

export interface LocalizationFormProps extends IssueProps {
  language: Asset;
  nodeSettings: NodeEditorSettings;
  updateLocalizations(languageCode: string, localizations: any[]): void;
  onClose(canceled: boolean): void;
  helpArticles: { [key: string]: string };
}

export const NAME_PROPERTY: Asset = {
  name: titleCase(ContactProperties.Name),
  id: ContactProperties.Name,
  type: AssetType.ContactProperty
};

export const CHANNEL_PROPERTY: Asset = {
  name: titleCase(ContactProperties.Channel),
  id: ContactProperties.Channel,
  type: AssetType.ContactProperty
};

export const LANGUAGE_PROPERTY: Asset = {
  name: titleCase(ContactProperties.Language),
  id: ContactProperties.Language,
  type: AssetType.ContactProperty
};

export const STATUS_PROPERTY: Asset = {
  name: titleCase(ContactProperties.Status),
  id: ContactProperties.Status,
  type: AssetType.ContactProperty
};
