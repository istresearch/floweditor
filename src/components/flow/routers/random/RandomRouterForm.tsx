import { default as bindCallbacks } from 'auto-bind/react';
import Dialog, { ButtonSet } from 'components/dialog/Dialog';
import { hasErrors, renderIssues } from 'components/flow/actions/helpers';
import { RouterFormProps } from 'components/flow/props';
import { createResultNameInput } from 'components/flow/routers/widgets';
import SelectElement, { SelectOption } from 'components/form/select/SelectElement';
import TextInputElement from 'components/form/textinput/TextInputElement';
import TypeList from 'components/nodeeditor/TypeList';
import { fakePropType } from 'config/ConfigProvider';
import { Category } from 'flowTypes';
import * as React from 'react';
import { FormState, mergeForm, SelectOptionEntry, StringEntry } from 'store/nodeEditor';
import { Alphanumeric, StartIsNonNumeric, validate } from 'store/validators';

import { BUCKET_OPTIONS, fillOutCategories, nodeToState, stateToNode } from './helpers';
import styles from './RandomRouterForm.module.scss';
import i18n from 'config/i18n';
import { TembaSelectStyle } from 'temba/TembaSelect';

// TODO: Remove use of Function
// tslint:disable:ban-types
export enum InputToFocus {
  args = 'args',
  min = 'min',
  max = 'max',
  exit = 'exit'
}

export interface RandomRouterFormState extends FormState {
  bucketChoice: SelectOptionEntry;
  resultName: StringEntry;
  categories: Category[];
}

export const leadInSpecId = 'lead-in';

export default class RandomRouterForm extends React.Component<
  RouterFormProps,
  RandomRouterFormState
> {
  constructor(props: RouterFormProps) {
    super(props);

    this.state = nodeToState(this.props.nodeSettings);

    bindCallbacks(this, {
      include: [/^on/, /^handle/]
    });
  }

  public static contextTypes = {
    assetService: fakePropType
  };

  private handleUpdateResultName(value: string): void {
    const resultName = validate(i18n.t('forms.result_name', 'Result Name'), value, [
      Alphanumeric,
      StartIsNonNumeric
    ]);
    this.setState({
      resultName,
      valid: this.state.valid && !hasErrors(resultName)
    });
  }

  private handleBucketsChanged(selected: SelectOption): boolean {
    // create new exits if needed

    const count = parseInt(selected.value, 10);

    let categories = this.state.categories.concat([]);

    // prune off if we have too many
    categories = categories.slice(0, count);

    // add any that we still need
    categories = fillOutCategories(categories, count);

    const updates: Partial<RandomRouterFormState> = {
      bucketChoice: { value: selected }
    };

    const updated = mergeForm(this.state, updates);
    this.setState({ ...updated, categories });

    return updated.valid;
  }

  public handleSave(): void {
    // don't proceed without our categories filled
    if (this.state.categories.find(cat => cat.name.trim().length === 0)) {
      return;
    }

    this.props.updateRouter(stateToNode(this.props.nodeSettings, this.state));
    this.props.onClose(false);
  }

  private getButtons(): ButtonSet {
    return {
      primary: { name: i18n.t('buttons.ok', 'Ok'), onClick: this.handleSave },
      secondary: {
        name: i18n.t('buttons.cancel', 'Cancel'),
        onClick: () => this.props.onClose(true)
      }
    };
  }

  private handleBucketNameChanged(category: Category, value: string): void {
    const categories = this.state.categories;
    categories.find((cat: Category) => cat.uuid === category.uuid).name = value;
    this.setState({ categories });
  }

  private renderBucketNames(): any {
    return this.state.categories.map((cat: Category) => (
      <TextInputElement
        key={cat.uuid}
        __className={styles.bucket_name}
        name={cat.uuid}
        entry={{
          value: cat.name,
          validationFailures:
            cat.name.trim().length === 0 ? [{ message: 'Bucket name is required' }] : []
        }}
        onChange={(value: string) => {
          this.handleBucketNameChanged(cat, value);
        }}
      />
    ));
  }

  public renderEdit(): JSX.Element {
    const typeConfig = this.props.typeConfig;

    const OPTIONS = BUCKET_OPTIONS.concat([]);
    if (BUCKET_OPTIONS.indexOf(this.state.bucketChoice.value) === -1) {
      OPTIONS.push(this.state.bucketChoice.value);
    }

    return (
      <Dialog title={typeConfig.name} headerClass={typeConfig.type} buttons={this.getButtons()}>
        <TypeList __className="" initialType={typeConfig} onChange={this.props.onTypeChange} />
        <div className={styles.options}>
          <div className={styles.lead_in}>Split them randomly into one of</div>
          <div className={styles.bucket_select}>
            <SelectElement
              key="buckets_select"
              style={TembaSelectStyle.small}
              name={i18n.t('forms.buckets', 'Buckets')}
              entry={this.state.bucketChoice}
              onChange={this.handleBucketsChanged}
              options={OPTIONS}
            />
          </div>
        </div>
        <div className={styles.bucket_list}>{this.renderBucketNames()}</div>
        {createResultNameInput(this.state.resultName, this.handleUpdateResultName)}
        {renderIssues(this.props)}
      </Dialog>
    );
  }

  public render(): JSX.Element {
    return this.renderEdit();
  }
}
