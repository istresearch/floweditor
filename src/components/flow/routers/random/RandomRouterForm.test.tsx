import { RouterFormProps } from 'components/flow/props';
import { getTypeConfig } from 'config';
import { Types } from 'config/interfaces';
import * as React from 'react';
import { fireEvent, render, fireTembaSelect, getByDisplayValue, getByTitle } from 'test/utils';
import { composeComponentTestUtils, mock } from 'testUtils';
import {
  createRandomNode,
  createRenderNode,
  createSendMsgAction,
  getRouterFormProps
} from 'testUtils/assetCreators';
import * as utils from 'utils';

import RandomRouterForm from './RandomRouterForm';

mock(utils, 'createUUID', utils.seededUUIDs());

const { setup } = composeComponentTestUtils<RouterFormProps>(
  RandomRouterForm,
  getRouterFormProps(createRandomNode(2))
);

describe(RandomRouterForm.name, () => {
  it('should render', () => {
    const { wrapper } = setup(true);
    expect(wrapper).toMatchSnapshot();
  });

  it('should initialize existing random', () => {
    const props: RouterFormProps = {
      nodeSettings: {
        originalNode: createRandomNode(5)
      },
      typeConfig: getTypeConfig(Types.split_by_random),
      assetStore: null,
      updateRouter: jest.fn(),
      onTypeChange: jest.fn(),
      onClose: jest.fn(),
      issues: [],
      helpArticles: {}
    };

    const { baseElement } = render(<RandomRouterForm {...props} />);
    expect(baseElement).toMatchSnapshot();
  });

  it('should remove exits when shrinking', () => {
    const props: RouterFormProps = {
      nodeSettings: {
        originalNode: createRandomNode(5)
      },
      typeConfig: getTypeConfig(Types.split_by_random),
      assetStore: null,
      updateRouter: jest.fn(),
      onTypeChange: jest.fn(),
      onClose: jest.fn(),
      helpArticles: {},
      issues: []
    };

    const {
      baseElement,
      getAllByTestId,
      getByText,
      getByDisplayValue,
      getByTitle,
      getByTestId
    } = render(<RandomRouterForm {...props} />);

    // we start off with five input boxes for our buckets
    expect(baseElement.querySelectorAll('input').length).toEqual(5);

    // choose 3 buckets
    fireTembaSelect(getByTestId('temba_select_buckets'), '3');

    // now we should only have three input buckets
    expect(baseElement.querySelectorAll('input').length).toEqual(3);
    expect(baseElement).toMatchSnapshot();

    // now lets save our form
    fireEvent.click(getByText('Ok'));

    expect(props.updateRouter).toMatchCallSnapshot();
  });

  it('should convert from a non-random node', () => {
    const { wrapper, instance, props } = setup(true, {
      nodeSettings: {
        $set: {
          originalAction: createSendMsgAction({ text: '' }),
          originalNode: createRenderNode({
            actions: [createSendMsgAction({ text: 'A message' })],
            exits: [{ uuid: utils.createUUID() }],
            ui: {
              type: Types.execute_actions,
              position: { left: 100, top: 100 }
            }
          })
        }
      }
    });

    instance.handleSave();

    // our orginal node should still only have one exit
    expect(props.nodeSettings.originalNode.node.exits.length).toBe(1);

    expect(wrapper).toMatchSnapshot();
    expect(props.updateRouter).toMatchCallSnapshot();
    expect(props.nodeSettings.originalNode).toMatchSnapshot();
  });
});
