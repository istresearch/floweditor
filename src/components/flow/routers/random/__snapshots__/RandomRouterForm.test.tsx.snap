// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RandomRouterForm should convert from a non-random node 1`] = `
<Dialog
  buttons={
    Object {
      "primary": Object {
        "name": "Ok",
        "onClick": [Function],
      },
      "secondary": Object {
        "name": "Cancel",
        "onClick": [Function],
      },
    }
  }
  headerClass="split_by_random"
  title="Split Randomly"
>
  <TypeList
    __className=""
    initialType={
      Object {
        "description": "Split by random chance",
        "form": [Function],
        "localization": [Function],
        "localizeableKeys": Array [
          "exits",
        ],
        "name": "Split Randomly",
        "type": "split_by_random",
      }
    }
    onChange={[MockFunction]}
  />
  <div
    className="options"
  >
    <div
      className="lead_in"
    >
      Split them randomly into one of
    </div>
    <div
      className="bucket_select"
    >
      <SelectElement
        entry={
          Object {
            "value": Object {
              "name": "2 buckets",
              "value": "2",
            },
          }
        }
        key="buckets_select"
        name="Buckets"
        onChange={[Function]}
        options={
          Array [
            Object {
              "name": "2 buckets",
              "value": "2",
            },
            Object {
              "name": "3 buckets",
              "value": "3",
            },
            Object {
              "name": "4 buckets",
              "value": "4",
            },
            Object {
              "name": "5 buckets",
              "value": "5",
            },
            Object {
              "name": "6 buckets",
              "value": "6",
            },
            Object {
              "name": "7 buckets",
              "value": "7",
            },
            Object {
              "name": "8 buckets",
              "value": "8",
            },
            Object {
              "name": "9 buckets",
              "value": "9",
            },
            Object {
              "name": "10 buckets",
              "value": "10",
            },
          ]
        }
        style="small"
      />
    </div>
  </div>
  <div
    className="bucket_list"
  >
    <TextInputElement
      __className="bucket_name"
      entry={
        Object {
          "value": "Bucket 1",
        }
      }
      key="306dee46-46c7-4a8a-8a33-ab2b7e42b02e"
      name="306dee46-46c7-4a8a-8a33-ab2b7e42b02e"
      onChange={[Function]}
    />
    <TextInputElement
      __className="bucket_name"
      entry={
        Object {
          "value": "Bucket 2",
        }
      }
      key="90e20a49-6ecc-4b0b-97bf-bcbd37069b63"
      name="90e20a49-6ecc-4b0b-97bf-bcbd37069b63"
      onChange={[Function]}
    />
  </div>
  <OptionalTextInput
    helpText={
      <Trans
        i18nKey="forms.result_name_help"
        values={
          Object {
            "resultFormat": "@results",
          }
        }
      >
        By naming the result, you can reference it later using [[resultFormat]]
      </Trans>
    }
    maxLength={64}
    name="Result Name"
    onChange={[Function]}
    toggleText="Save as.."
    value={
      Object {
        "value": "",
      }
    }
  />
</Dialog>
`;

exports[`RandomRouterForm should convert from a non-random node 2`] = `
Array [
  Object {
    "inboundConnections": Object {},
    "node": Object {
      "actions": Array [],
      "exits": Array [
        Object {
          "destination_uuid": null,
          "uuid": "b5292cef-6113-4066-80f5-371178abf28b",
        },
        Object {
          "destination_uuid": null,
          "uuid": "f627a69c-a227-451f-9520-ce969f02162e",
        },
      ],
      "router": Object {
        "categories": Array [
          Object {
            "exit_uuid": "b5292cef-6113-4066-80f5-371178abf28b",
            "name": "Bucket 1",
            "uuid": "306dee46-46c7-4a8a-8a33-ab2b7e42b02e",
          },
          Object {
            "exit_uuid": "f627a69c-a227-451f-9520-ce969f02162e",
            "name": "Bucket 2",
            "uuid": "90e20a49-6ecc-4b0b-97bf-bcbd37069b63",
          },
        ],
        "type": "random",
      },
      "uuid": "c99dcf07-2ad6-4730-9e99-810a39f5a58d",
    },
    "ui": Object {
      "config": null,
      "position": null,
      "type": "split_by_random",
    },
  },
]
`;

exports[`RandomRouterForm should convert from a non-random node 3`] = `
Object {
  "inboundConnections": Object {},
  "node": Object {
    "actions": Array [
      Object {
        "all_urns": false,
        "text": "A message",
        "type": "send_msg",
        "uuid": "38282c32-100e-4376-af57-7f6a36ddafba",
      },
    ],
    "exits": Array [
      Object {
        "uuid": "b5a6cc81-02ba-4931-b4be-a0d7204be029",
      },
    ],
    "uuid": "c99dcf07-2ad6-4730-9e99-810a39f5a58d",
  },
  "ui": Object {
    "position": Object {
      "left": 100,
      "top": 100,
    },
    "type": "execute_actions",
  },
}
`;

exports[`RandomRouterForm should initialize existing random 1`] = `
<body>
  <div>
    <div
      class="dialog"
    >
      <div
        class="header split_by_random"
      >
        <div
          class="title_container"
        >
          <div
            class="title"
          >
            Split Randomly
          </div>
          <div
            class="subtitle"
          />
        </div>
      </div>
      <div
        class="content"
      >
        <div
          class=" type_list"
        >
          <p>
            When a contact arrives at this point in your flow...
          </p>
          <div>
            <div
              class="normal "
            >
              <temba-select
                data-testid="temba_select_type"
                errors="[]"
                expressions=""
                name="type"
                namekey="description"
                searchable="true"
                valuekey="type"
                values="[{\\"type\\":\\"split_by_random\\",\\"name\\":\\"Split Randomly\\",\\"description\\":\\"Split by random chance\\",\\"localizeableKeys\\":[\\"exits\\"]}]"
              />
            </div>
          </div>
        </div>
        <div
          class="options"
        >
          <div
            class="lead_in"
          >
            Split them randomly into one of
          </div>
          <div
            class="bucket_select"
          >
            <div
              class="ele"
            >
              <div
                class="small "
              >
                <temba-select
                  data-testid="temba_select_buckets"
                  errors="[]"
                  expressions=""
                  name="Buckets"
                  namekey="name"
                  valuekey="value"
                  values="[{\\"value\\":\\"5\\",\\"name\\":\\"5 buckets\\"}]"
                />
              </div>
            </div>
          </div>
        </div>
        <div
          class="bucket_list"
        >
          <div
            class="ele bucket_name"
          >
            <div
              class="wrapper normal"
            >
              <input
                data-testid="e67ce051-f3e4-4c42-807b-e393df5fa382"
                name="e67ce051-f3e4-4c42-807b-e393df5fa382"
                value="Bucket 1"
              />
            </div>
          </div>
          <div
            class="ele bucket_name"
          >
            <div
              class="wrapper normal"
            >
              <input
                data-testid="e0bc3776-792f-49d2-b4be-48be2cf8159a"
                name="e0bc3776-792f-49d2-b4be-48be2cf8159a"
                value="Bucket 2"
              />
            </div>
          </div>
          <div
            class="ele bucket_name"
          >
            <div
              class="wrapper normal"
            >
              <input
                data-testid="6c22884b-0e35-4206-982e-18320691eda9"
                name="6c22884b-0e35-4206-982e-18320691eda9"
                value="Bucket 3"
              />
            </div>
          </div>
          <div
            class="ele bucket_name"
          >
            <div
              class="wrapper normal"
            >
              <input
                data-testid="a62e8284-49f8-4292-adbc-d5f6607f85c5"
                name="a62e8284-49f8-4292-adbc-d5f6607f85c5"
                value="Bucket 4"
              />
            </div>
          </div>
          <div
            class="ele bucket_name"
          >
            <div
              class="wrapper normal"
            >
              <input
                data-testid="0c8c9239-9b59-473a-a229-d8a9693be270"
                name="0c8c9239-9b59-473a-a229-d8a9693be270"
                value="Bucket 5"
              />
            </div>
          </div>
        </div>
        <div
          class="optional_text_input"
        >
          <span
            class="toggle_link"
            data-spec="toggle-link"
            data-testid="toggle-link"
          >
            Save as..
          </span>
        </div>
      </div>
      <div
        class="footer"
      >
        <div
          class="buttons"
        >
          <div
            class="right_buttons"
          >
            <div
              class="btn secondary  "
              style="margin-left: 0px; margin-top: 0px;"
            >
              Cancel
            </div>
            <div
              class="btn primary  "
              style="margin-left: 10px; margin-top: 0px;"
            >
              Ok
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`RandomRouterForm should remove exits when shrinking 1`] = `
<body>
  <div>
    <div
      class="dialog"
    >
      <div
        class="header split_by_random"
      >
        <div
          class="title_container"
        >
          <div
            class="title"
          >
            Split Randomly
          </div>
          <div
            class="subtitle"
          />
        </div>
      </div>
      <div
        class="content"
      >
        <div
          class=" type_list"
        >
          <p>
            When a contact arrives at this point in your flow...
          </p>
          <div>
            <div
              class="normal "
            >
              <temba-select
                data-testid="temba_select_type"
                errors="[]"
                expressions=""
                name="type"
                namekey="description"
                searchable="true"
                valuekey="type"
                values="[{\\"type\\":\\"split_by_random\\",\\"name\\":\\"Split Randomly\\",\\"description\\":\\"Split by random chance\\",\\"localizeableKeys\\":[\\"exits\\"]}]"
              />
            </div>
          </div>
        </div>
        <div
          class="options"
        >
          <div
            class="lead_in"
          >
            Split them randomly into one of
          </div>
          <div
            class="bucket_select"
          >
            <div
              class="ele"
            >
              <div
                class="small "
              >
                <temba-select
                  data-testid="temba_select_buckets"
                  errors="[]"
                  expressions=""
                  name="Buckets"
                  namekey="name"
                  valuekey="value"
                  values="[{\\"value\\":\\"3\\",\\"name\\":\\"3 buckets\\"}]"
                />
              </div>
            </div>
          </div>
        </div>
        <div
          class="bucket_list"
        >
          <div
            class="ele bucket_name"
          >
            <div
              class="wrapper normal"
            >
              <input
                data-testid="e81b3f1a-0863-4418-8a72-9c9f8aa72b7c"
                name="e81b3f1a-0863-4418-8a72-9c9f8aa72b7c"
                value="Bucket 1"
              />
            </div>
          </div>
          <div
            class="ele bucket_name"
          >
            <div
              class="wrapper normal"
            >
              <input
                data-testid="1888ae1f-2518-4727-98e2-731674e1760e"
                name="1888ae1f-2518-4727-98e2-731674e1760e"
                value="Bucket 2"
              />
            </div>
          </div>
          <div
            class="ele bucket_name"
          >
            <div
              class="wrapper normal"
            >
              <input
                data-testid="dd57500e-a8e8-4a98-8b5b-a33637523a8d"
                name="dd57500e-a8e8-4a98-8b5b-a33637523a8d"
                value="Bucket 3"
              />
            </div>
          </div>
        </div>
        <div
          class="optional_text_input"
        >
          <span
            class="toggle_link"
            data-spec="toggle-link"
            data-testid="toggle-link"
          >
            Save as..
          </span>
        </div>
      </div>
      <div
        class="footer"
      >
        <div
          class="buttons"
        >
          <div
            class="right_buttons"
          >
            <div
              class="btn secondary  "
              style="margin-left: 0px; margin-top: 0px;"
            >
              Cancel
            </div>
            <div
              class="btn primary  "
              style="margin-left: 10px; margin-top: 0px;"
            >
              Ok
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`RandomRouterForm should remove exits when shrinking 2`] = `
Array [
  Object {
    "inboundConnections": Object {},
    "node": Object {
      "actions": Array [],
      "exits": Array [
        Object {
          "destination_uuid": null,
          "uuid": "6b59e863-2e14-442b-9987-aaa284871b8e",
        },
        Object {
          "destination_uuid": null,
          "uuid": "00dd2423-0a34-4ee2-b97d-839fb2d593ef",
        },
        Object {
          "destination_uuid": null,
          "uuid": "c5722340-f7e1-4d99-ac5e-3a6e59845111",
        },
      ],
      "router": Object {
        "categories": Array [
          Object {
            "exit_uuid": "6b59e863-2e14-442b-9987-aaa284871b8e",
            "name": "Bucket 1",
            "uuid": "e81b3f1a-0863-4418-8a72-9c9f8aa72b7c",
          },
          Object {
            "exit_uuid": "00dd2423-0a34-4ee2-b97d-839fb2d593ef",
            "name": "Bucket 2",
            "uuid": "1888ae1f-2518-4727-98e2-731674e1760e",
          },
          Object {
            "exit_uuid": "c5722340-f7e1-4d99-ac5e-3a6e59845111",
            "name": "Bucket 3",
            "uuid": "dd57500e-a8e8-4a98-8b5b-a33637523a8d",
          },
        ],
        "type": "random",
      },
      "uuid": "96249bdc-cb58-4489-a188-0eeb6686a8c9",
    },
    "ui": Object {
      "config": null,
      "position": null,
      "type": "split_by_random",
    },
  },
]
`;

exports[`RandomRouterForm should render 1`] = `
<Dialog
  buttons={
    Object {
      "primary": Object {
        "name": "Ok",
        "onClick": [Function],
      },
      "secondary": Object {
        "name": "Cancel",
        "onClick": [Function],
      },
    }
  }
  headerClass="split_by_random"
  title="Split Randomly"
>
  <TypeList
    __className=""
    initialType={
      Object {
        "description": "Split by random chance",
        "form": [Function],
        "localization": [Function],
        "localizeableKeys": Array [
          "exits",
        ],
        "name": "Split Randomly",
        "type": "split_by_random",
      }
    }
    onChange={[MockFunction]}
  />
  <div
    className="options"
  >
    <div
      className="lead_in"
    >
      Split them randomly into one of
    </div>
    <div
      className="bucket_select"
    >
      <SelectElement
        entry={
          Object {
            "value": Object {
              "name": "2 buckets",
              "value": "2",
            },
          }
        }
        key="buckets_select"
        name="Buckets"
        onChange={[Function]}
        options={
          Array [
            Object {
              "name": "2 buckets",
              "value": "2",
            },
            Object {
              "name": "3 buckets",
              "value": "3",
            },
            Object {
              "name": "4 buckets",
              "value": "4",
            },
            Object {
              "name": "5 buckets",
              "value": "5",
            },
            Object {
              "name": "6 buckets",
              "value": "6",
            },
            Object {
              "name": "7 buckets",
              "value": "7",
            },
            Object {
              "name": "8 buckets",
              "value": "8",
            },
            Object {
              "name": "9 buckets",
              "value": "9",
            },
            Object {
              "name": "10 buckets",
              "value": "10",
            },
          ]
        }
        style="small"
      />
    </div>
  </div>
  <div
    className="bucket_list"
  >
    <TextInputElement
      __className="bucket_name"
      entry={
        Object {
          "value": "Bucket 1",
        }
      }
      key="1dce2b34-9aab-4e20-81c4-3f0408dcb671"
      name="1dce2b34-9aab-4e20-81c4-3f0408dcb671"
      onChange={[Function]}
    />
    <TextInputElement
      __className="bucket_name"
      entry={
        Object {
          "value": "Bucket 2",
        }
      }
      key="763e4844-3e1b-407a-a1b5-5fdfcd308b41"
      name="763e4844-3e1b-407a-a1b5-5fdfcd308b41"
      onChange={[Function]}
    />
  </div>
  <OptionalTextInput
    helpText={
      <Trans
        i18nKey="forms.result_name_help"
        values={
          Object {
            "resultFormat": "@results",
          }
        }
      >
        By naming the result, you can reference it later using [[resultFormat]]
      </Trans>
    }
    maxLength={64}
    name="Result Name"
    onChange={[Function]}
    toggleText="Save as.."
    value={
      Object {
        "value": "",
      }
    }
  />
</Dialog>
`;
